### ISSUE TITLE
Go Programming

### ISSUE THREAD

**Issue by SpheMakh** [ratt-ru/reading-list#12]
This looks pretty cool, if you keen on learning a new language: http://www.golang-book.com/books/intro

**Comment by sphemakh** [comment]
If you think reading is overrated : https://www.youtube.com/watch?v=CF9S4QZuV30

**Comment by gigjozsa** [comment]
This is indeed a nice introduction, but why go? It looks like some 
mixture of c and python ideas.

On 12/06/2015 00:46, Sphesihle Makhathini wrote:

> This looks pretty cool, if you keen on learning a new language: 
> http://www.golang-book.com/books/intro
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/ratt-ru/reading-list/issues/12.

## 

_Dr. Gyula I. G. Józsa_
Astronomer
Rhodes University  http://www.ru.ac.za/,RARG  http://www.ru.ac.za/physicsandelectronics/radioastronomycentre/,RATT  http://www.scifac.ru.ac.za/ratt/, PO Box 94, Grahamstown, 6140, South Africa andSKA South Africa  http://www.ska.ac.za/, The Park, Park Road, Pinelands, 7405, South Africa
Email:jozsa@ska.ac.za  mailto:jozsa@ska.ac.za    Phone: +27 (0)21 506 7369   Fax: +27 (0)21 506 7375

**Comment by gigjozsa** [comment]
c with garbage control, some objective elements, some pythonic types, 
plus, I forgot before, concurrent elements (also available as omp for c)?

On 12/06/2015 00:46, Sphesihle Makhathini wrote:

> This looks pretty cool, if you keen on learning a new language: 
> http://www.golang-book.com/books/intro
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/ratt-ru/reading-list/issues/12.

## 

_Dr. Gyula I. G. Józsa_
Astronomer
Rhodes University  http://www.ru.ac.za/,RARG  http://www.ru.ac.za/physicsandelectronics/radioastronomycentre/,RATT  http://www.scifac.ru.ac.za/ratt/, PO Box 94, Grahamstown, 6140, South Africa andSKA South Africa  http://www.ska.ac.za/, The Park, Park Road, Pinelands, 7405, South Africa
Email:jozsa@ska.ac.za  mailto:jozsa@ska.ac.za    Phone: +27 (0)21 506 7369   Fax: +27 (0)21 506 7375

**Comment by gijzelaerr** [comment]
O I love new languages.

Go is still quite young, but is rapidly gaining popularity for writing all kind of funky new things like docker and network services. Problem is that is doesn't have proper numerical libraries, unless you are willing to interface with C.

if you are investigating languages since you have spare time, you are just curious or just like to learn something new I also suggest:

JuliaLang - the scientific compute language, build for numerical shizzle, looks a lot like Python but faster and without the awkwardness. It sort of is what Python should have been and interfaces very well with Python, Numpy and Scipy (even better than PyPy does):
http://julialang.org/

For something completely different and the next level mind exercise I recommend Haskell, a purely functional high level language which is awesome but a bit of a intellectual challenge:
https://www.haskell.org/

The type system of haskell is particularly very interesting and I'll buy you a beer when you understand monads. Since types seem to be such a good idea in some cases, for Python 3.5 the developers are introducing optional type checking to the language, which is backported as a module:

https://pypi.python.org/pypi/typing/3.5.0b1

Python 3.5 is going to be interesting also for multiple reasons, for example a matrix multiplication operator

https://www.python.org/dev/peps/pep-0465/

And last but not least allowing parallel build of C extensions, which will greatly speed up compilation of stuff like numpy on multicore systems:

https://docs.python.org/dev/whatsnew/3.5.html#distutils

So i'm drifting a bit here, but now we are talking about languages anyway, scientists seem to be a bit stuck in the Python 2.7 landscape, which was released about 5 years ago. So if you want to satisfy your language fetish go make your code python 3 compatible so we can ditch python 2 at some point and never have to look back:

http://python-future.org/compatible_idioms.html

**Comment by gigjozsa** [comment]
Thanks, especially for the remark about Python 3. This would, however, 
require a common effort, wouldn't it.

Apart from that, unfortunately, while I, too, love languages, there's 
currently not much spare time apart from when I wake up in the middle of 
the night and try to distract myself.

On 12/06/2015 10:27, Gijs Molenaar wrote:

> O I love new languages.
> 
> Go is still quite young, but is rapidly gaining popularity for writing 
> all kind of funky new things like docker and network services. Problem 
> is that is doesn't have proper numerical libraries, unless you are 
> willing to interface with C.
> 
> if you are investigating languages since you have spare time, you are 
> just curious or just like to learn something new I also suggest:
> 
> JuliaLang - the scientific compute language, build for numerical 
> shizzle, looks a lot like Python but faster and without the 
> awkwardness. It sort of is what Python should have been and interfaces 
> very well with Python, Numpy and Scipy (even better than PyPy does):
> http://julialang.org/
> 
> For something completely different and the next level mind exercise I 
> recommend Haskell, a purely functional high level language which is 
> awesome but a bit of a intellectual challenge:
> https://www.haskell.org/
> 
> The type system of haskell is particularly very interesting and I'll 
> buy you a beer when you understand monads. Since types seem to be such 
> a good idea in some cases, for Python 3.5 the developers are 
> introducing optional type checking to the language, which is 
> backported as a module:
> 
> https://pypi.python.org/pypi/typing/3.5.0b1
> 
> Python 3.5 is going to be interesting also for multiple reasons, for 
> example a matrix multiplication operator
> 
> https://www.python.org/dev/peps/pep-0465/
> 
> And last but not least allowing parallel build of C extensions, which 
> will greatly speed up compilation of stuff like numpy on multicore 
> systems:
> 
> https://docs.python.org/dev/whatsnew/3.5.html#distutils
> 
> So i'm drifting a bit here, but now we are talking about languages 
> anyway, scientists seem to be a bit stuck in the Python 2.7 landscape, 
> which was released about 5 years ago. So if you want to satisfy your 
> language fetish go make your code python 3 compatible so we can ditch 
> python 2 at some point and never have to look back:
> 
> http://python-future.org/compatible_idioms.html
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/ratt-ru/reading-list/issues/12#issuecomment-111411888.

## 

_Dr. Gyula I. G. Józsa_
Astronomer
Rhodes University  http://www.ru.ac.za/,RARG  http://www.ru.ac.za/physicsandelectronics/radioastronomycentre/,RATT  http://www.scifac.ru.ac.za/ratt/, PO Box 94, Grahamstown, 6140, South Africa andSKA South Africa  http://www.ska.ac.za/, The Park, Park Road, Pinelands, 7405, South Africa
Email:jozsa@ska.ac.za  mailto:jozsa@ska.ac.za    Phone: +27 (0)21 506 7369   Fax: +27 (0)21 506 7375

**Comment by gijzelaerr** [comment]
That must be a very good middle-of-the-night distraction if it has priority
over learning a new programming language.

2015-06-12 10:43 GMT+02:00 Gyula Jozsa (Josh) notifications@github.com:

> Thanks, especially for the remark about Python 3. This would, however,
> require a common effort, wouldn't it.
> 
> Apart from that, unfortunately, while I, too, love languages, there's
> currently not much spare time apart from when I wake up in the middle of
> the night and try to distract myself.
> 
> On 12/06/2015 10:27, Gijs Molenaar wrote:
> 
> > O I love new languages.
> > 
> > Go is still quite young, but is rapidly gaining popularity for writing
> > all kind of funky new things like docker and network services. Problem
> > is that is doesn't have proper numerical libraries, unless you are
> > willing to interface with C.
> > 
> > if you are investigating languages since you have spare time, you are
> > just curious or just like to learn something new I also suggest:
> > 
> > JuliaLang - the scientific compute language, build for numerical
> > shizzle, looks a lot like Python but faster and without the
> > awkwardness. It sort of is what Python should have been and interfaces
> > very well with Python, Numpy and Scipy (even better than PyPy does):
> > http://julialang.org/
> > 
> > For something completely different and the next level mind exercise I
> > recommend Haskell, a purely functional high level language which is
> > awesome but a bit of a intellectual challenge:
> > https://www.haskell.org/
> > 
> > The type system of haskell is particularly very interesting and I'll
> > buy you a beer when you understand monads. Since types seem to be such
> > a good idea in some cases, for Python 3.5 the developers are
> > introducing optional type checking to the language, which is
> > backported as a module:
> > 
> > https://pypi.python.org/pypi/typing/3.5.0b1
> > 
> > Python 3.5 is going to be interesting also for multiple reasons, for
> > example a matrix multiplication operator
> > 
> > https://www.python.org/dev/peps/pep-0465/
> > 
> > And last but not least allowing parallel build of C extensions, which
> > will greatly speed up compilation of stuff like numpy on multicore
> > systems:
> > 
> > https://docs.python.org/dev/whatsnew/3.5.html#distutils
> > 
> > So i'm drifting a bit here, but now we are talking about languages
> > anyway, scientists seem to be a bit stuck in the Python 2.7 landscape,
> > which was released about 5 years ago. So if you want to satisfy your
> > language fetish go make your code python 3 compatible so we can ditch
> > python 2 at some point and never have to look back:
> > 
> > http://python-future.org/compatible_idioms.html
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > <
> > https://github.com/ratt-ru/reading-list/issues/12#issuecomment-111411888>.
> 
> ## 
> 
> _Dr. Gyula I. G. Józsa_
> Astronomer
> Rhodes University http://www.ru.ac.za/,RARG <
> http://www.ru.ac.za/physicsandelectronics/radioastronomycentre/>,RATT <
> http://www.scifac.ru.ac.za/ratt/>, PO Box 94, Grahamstown, 6140, South
> Africa andSKA South Africa http://www.ska.ac.za/, The Park, Park Road,
> Pinelands, 7405, South Africa
> Email:jozsa@ska.ac.za mailto:jozsa@ska.ac.za Phone: +27 (0)21 506 7369
> Fax: +27 (0)21 506 7375
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ratt-ru/reading-list/issues/12#issuecomment-111415935
> .

## 

Gijs Molenaar
http://pythonic.nl

**Comment by o-smirnov** [comment]
The new MFS algorithm paper comes with a Julia implementation. We need to try running it: http://arxiv.org/abs/1504.06847

We also need to find a person named Mercedes to include on the author list.

**Comment by gigjozsa** [comment]
Truly, and such I learned about go last night.

On 12/06/2015 10:46, Gijs Molenaar wrote:

> That must be a very good middle-of-the-night distraction if it has 
> priority
> over learning a new programming language.
> 
> 2015-06-12 10:43 GMT+02:00 Gyula Jozsa (Josh) notifications@github.com:
> 
> > Thanks, especially for the remark about Python 3. This would, however,
> > require a common effort, wouldn't it.
> > 
> > Apart from that, unfortunately, while I, too, love languages, there's
> > currently not much spare time apart from when I wake up in the middle of
> > the night and try to distract myself.
> > 
> > On 12/06/2015 10:27, Gijs Molenaar wrote:
> > 
> > > O I love new languages.
> > > 
> > > Go is still quite young, but is rapidly gaining popularity for writing
> > > all kind of funky new things like docker and network services. Problem
> > > is that is doesn't have proper numerical libraries, unless you are
> > > willing to interface with C.
> > > 
> > > if you are investigating languages since you have spare time, you are
> > > just curious or just like to learn something new I also suggest:
> > > 
> > > JuliaLang - the scientific compute language, build for numerical
> > > shizzle, looks a lot like Python but faster and without the
> > > awkwardness. It sort of is what Python should have been and interfaces
> > > very well with Python, Numpy and Scipy (even better than PyPy does):
> > > http://julialang.org/
> > > 
> > > For something completely different and the next level mind exercise I
> > > recommend Haskell, a purely functional high level language which is
> > > awesome but a bit of a intellectual challenge:
> > > https://www.haskell.org/
> > > 
> > > The type system of haskell is particularly very interesting and I'll
> > > buy you a beer when you understand monads. Since types seem to be such
> > > a good idea in some cases, for Python 3.5 the developers are
> > > introducing optional type checking to the language, which is
> > > backported as a module:
> > > 
> > > https://pypi.python.org/pypi/typing/3.5.0b1
> > > 
> > > Python 3.5 is going to be interesting also for multiple reasons, for
> > > example a matrix multiplication operator
> > > 
> > > https://www.python.org/dev/peps/pep-0465/
> > > 
> > > And last but not least allowing parallel build of C extensions, which
> > > will greatly speed up compilation of stuff like numpy on multicore
> > > systems:
> > > 
> > > https://docs.python.org/dev/whatsnew/3.5.html#distutils
> > > 
> > > So i'm drifting a bit here, but now we are talking about languages
> > > anyway, scientists seem to be a bit stuck in the Python 2.7 landscape,
> > > which was released about 5 years ago. So if you want to satisfy your
> > > language fetish go make your code python 3 compatible so we can ditch
> > > python 2 at some point and never have to look back:
> > > 
> > > http://python-future.org/compatible_idioms.html
> > > 
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <
> > 
> > https://github.com/ratt-ru/reading-list/issues/12#issuecomment-111411888>.
> > 
> > 
> > ## 
> > 
> > _Dr. Gyula I. G. Józsa_
> > Astronomer
> > Rhodes University http://www.ru.ac.za/,RARG <
> > http://www.ru.ac.za/physicsandelectronics/radioastronomycentre/>,RATT <
> > http://www.scifac.ru.ac.za/ratt/>, PO Box 94, Grahamstown, 6140, South
> > Africa andSKA South Africa http://www.ska.ac.za/, The Park, Park Road,
> > Pinelands, 7405, South Africa
> > Email:jozsa@ska.ac.za mailto:jozsa@ska.ac.za Phone: +27 (0)21 506 7369
> > Fax: +27 (0)21 506 7375
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > 
> > https://github.com/ratt-ru/reading-list/issues/12#issuecomment-111415935
> > .
> 
> ## 
> 
> Gijs Molenaar
> http://pythonic.nl
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/ratt-ru/reading-list/issues/12#issuecomment-111416444.

## 

_Dr. Gyula I. G. Józsa_
Astronomer
Rhodes University  http://www.ru.ac.za/,RARG  http://www.ru.ac.za/physicsandelectronics/radioastronomycentre/,RATT  http://www.scifac.ru.ac.za/ratt/, PO Box 94, Grahamstown, 6140, South Africa andSKA South Africa  http://www.ska.ac.za/, The Park, Park Road, Pinelands, 7405, South Africa
Email:jozsa@ska.ac.za  mailto:jozsa@ska.ac.za    Phone: +27 (0)21 506 7369   Fax: +27 (0)21 506 7375

**Comment by cyriltasse** [comment]
@o-smirnov : you mention it's extremely fast. You run the simulation on a 256x256 image. Can this run on a 20.000x20.000 pixels image, and how long does it take?

**Comment by o-smirnov** [comment]
Well I'm only the 6th author on that so I only 1/6th mention... in fact I have trusted my co-authors on that...

**Comment by cyriltasse** [comment]
Looks cool anyway. But we should push CS people to really try to adress the "go big" problem.

**Comment by JSKenyon** [comment]
I agree completely. The absence of problems larger than 256x256 in the
literature is quite obvious. New methods are fantastic, but they do need to
be usable in the current context.

**Comment by o-smirnov** [comment]
It's parallel in frequency so would be interesting to test in in the cloud at some point...

**Comment by cyriltasse** [comment]
But at least intuitively there are ways to do.

**Comment by o-smirnov** [comment]
Yes, and given a GPU, e.g. PyMORESANE runs in a reasonable time on real-sized problems. If a code prototype can do 256x256 in MATLAB, it's already in reasonable territory, the rest is a matter of implementation and brute force.

**Comment by gijzelaerr** [comment]
and buying licenses

2015-06-12 11:38 GMT+02:00 Oleg Smirnov notifications@github.com:

> Yes, and given a GPU, e.g. PyMORESANE runs in a reasonable time on
> real-sized problems. If a code prototype can do 256x256 in MATLAB, it's
> already in reasonable territory, the rest is a matter of implementation and
> brute force.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ratt-ru/reading-list/issues/12#issuecomment-111429862
> .

## 

Gijs Molenaar
http://pythonic.nl
